src/screens/Exchange/ExchangeOffers.js

```diff
@@ -275,107 +274,107 @@ class ExchangeOffers extends React.Component<Props, State> {
-          const gasLimit = await calculateGasEstimate(transactionPayload);
-          txFeeInWei = gasPriceWei.mul(gasLimit);
-
-          transactionPayload = {
-            ...transactionPayload,
-            gasPrice: gasPriceWei,
-            gasLimit,
-            txSpeed: SPEED_TYPES.NORMAL,
-          };
-        }
-
-        transactionPayload = { ...transactionPayload, txFeeInWei };
-
-        const fiatCurrency = baseFiatCurrency || defaultFiatCurrency;
-        const feeSymbol = get(gasToken, 'symbol', ETH);
-        const feeDecimals = get(gasToken, 'decimals', 'ether');
-        const feeNumeric = utils.formatUnits(txFeeInWei.toString(), feeDecimals);
-        const feeInFiat = formatFiat(parseFloat(feeNumeric) * getRate(rates, feeSymbol, fiatCurrency), fiatCurrency);
-        const feeDisplayValue = formatTransactionFee(txFeeInWei, gasToken);
-        const isDisabled = !isEnoughBalanceForTransactionFee(balances, transactionPayload);
-
-        this.setState({
-          pressedTokenAllowanceId: '',
-          isEnableAssetModalVisible: true,
-          enableData: {
-            providerName,
-            assetSymbol,
-            assetIcon,
-            feeDisplayValue,
-            feeInFiat,
-            isDisabled,
-          },
-          enablePayload: { ...transactionPayload },
-        });
-      });
+    if (isEmpty(response)) {
+      this.setState({ pressedTokenAllowanceId: '' }); // reset set allowance button to be enabled
+      return;
+    }
+    setExecutingTransaction();
+    const {
+      payToAddress,
+      transactionObj: { data } = {},
+    } = response;
+
+    const assetToEnable = exchangeSupportedAssets.find(({ symbol }) => symbol === fromAssetCode) || {};
+    const { symbol: assetSymbol, iconUrl: assetIcon } = assetToEnable;
+    const providerName = getCryptoProviderName(providersMeta, provider);
+
+    let gasToken;
+    let txFeeInWei;
+    let transactionPayload = {
+      amount: 0,
+      to: payToAddress,
+      symbol: fromAssetCode,
+      contractAddress: fromAssetAddress || '',
+      decimals: parseInt(decimals, 10) || 18,
+      data,
+      extra: {
+        allowance: {
+          provider,
+          fromAssetCode,
+          toAssetCode,
+        },
+      },
+    };
+
+    if (isSmartAccount) {
+      ({ fee: txFeeInWei, gasToken } = await this.getSmartWalletTxFee(transactionPayload));
+      if (gasToken) {
+        transactionPayload = { ...transactionPayload, gasToken };
+      }
+    } else {
+      const gasPrice = gasInfo.gasPrice[SPEED_TYPES.NORMAL] || 0;
+      const gasPriceWei = utils.parseUnits(gasPrice.toString(), 'gwei');
+      const gasLimit = await calculateGasEstimate(transactionPayload);
+      txFeeInWei = gasPriceWei.mul(gasLimit);
+
+      transactionPayload = {
+        ...transactionPayload,
+        gasPrice: gasPriceWei,
+        gasLimit,
+        txSpeed: SPEED_TYPES.NORMAL,
+      };
+    }
```

Comment:
We no longer support Key based, this (and maybe I didn't notice elsewhere) can be optimized.